### PR TITLE
[BUG] Adding offset with nonzero month to DatetimeIndex (GH26258)

### LINKED ISSUE #26258: Adding offset to DatetimeIndex is broken
#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd
offset = pd.DateOffset(months=3, days=10)
dates = pd.date_range(start='01 Jan 2014', end='01 Jan 2017', freq='AS')
dates + offset
```

#### Problem description

ValueError is raised:

```python

Traceback (most recent call last):
  File "C:\mma\local\Anaconda3\envs\pd24\lib\site-packages\pandas\core\arrays\datetimelike.py", line 884, in _validate_frequency
    raise ValueError
ValueError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\mma\local\Anaconda3\envs\pd24\lib\site-packages\pandas\core\indexes\datetimelike.py", line 489, in __add__
    result = self._data.__add__(maybe_unwrap_index(other))
  File "C:\mma\local\Anaconda3\envs\pd24\lib\site-packages\pandas\core\arrays\datetimelike.py", line 1190, in __add__
    result = self._add_offset(other)
  File "C:\mma\local\Anaconda3\envs\pd24\lib\site-packages\pandas\core\arrays\datetimes.py", line 737, in _add_offset
    result = offset.apply_index(values)
  File "pandas/_libs/tslibs/offsets.pyx", line 116, in pandas._libs.tslibs.offsets.apply_index_wraps.wrapper
  File "C:\mma\local\Anaconda3\envs\pd24\lib\site-packages\pandas\tseries\offsets.py", line 278, in apply_index
    i = type(i)(shifted, freq=i.freq, dtype=i.dtype)
  File "C:\mma\local\Anaconda3\envs\pd24\lib\site-packages\pandas\core\arrays\datetimes.py", line 351, in __init__
    type(self)._validate_frequency(self, freq)
  File "C:\mma\local\Anaconda3\envs\pd24\lib\site-packages\pandas\core\arrays\datetimelike.py", line 897, in _validate_frequency
    .format(infer=inferred, passed=freq.freqstr))
ValueError: Inferred frequency AS-APR from passed values does not conform to passed frequency AS-JAN
```

#### Expected Output

This is what had been working through version 0.22.0 and was broken somewhere between 0.22 and 0.24:

```python
>>> dates + offset
DatetimeIndex(['2014-04-11', '2015-04-11', '2016-04-11', '2017-04-11'], dtype='datetime64[ns]', freq=None)
```
#### Workaround

```python

>>> pd.DatetimeIndex([date + offset for date in dates])
DatetimeIndex(['2014-04-11', '2015-04-11', '2016-04-11', '2017-04-11'], dtype='datetime64[ns]', freq=None)
```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.6.8.final.0
python-bits: 64
OS: Windows
OS-release: 7
machine: AMD64
processor: Intel64 Family 6 Model 58 Stepping 9, GenuineIntel
byteorder: little
LC_ALL: None
LANG: None
LOCALE: None.None

pandas: 0.24.2
pytest: 4.4.1
pip: 19.1
setuptools: 41.0.1
Cython: None
numpy: 1.16.3
scipy: None
pyarrow: None
xarray: None
IPython: None
sphinx: None
patsy: None
dateutil: 2.8.0
pytz: 2019.1
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: None
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml.etree: None
bs4: None
html5lib: None
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: None
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None
gcsfs: None

</details>

### PR DESCRIPTION
**Pull request by makbigc**
- [x] closes #26258 
- [x] 1 test added
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/tests/arithmetic/test_datetime64.py [hunk 1]
```diff
@@ -1370,6 +1370,14 @@ def test_dt64arr_add_sub_DateOffset(self, box_with_array):
...
         tm.assert_equal(result2, exp)
 
+        # GH 26258
+        date = date_range(start='01 Jan 2014', end='01 Jan 2017', freq='AS')
+        offset = pd.DateOffset(months=3, days=10)
+        result = date + offset
+        exp = DatetimeIndex([Timestamp('2014-04-11'), Timestamp('2015-04-11'),
+                             Timestamp('2016-04-11'), Timestamp('2017-04-11')])
+        tm.assert_equal(result, exp)
+
```
Comment (gfyoung) [inline pandas/tests/arithmetic/test_datetime64.py]: Make this a separate test.  We're moving towards smaller, more modular tests.

File: pandas/tests/arithmetic/test_datetime64.py [hunk 2]
```diff
@@ -1370,6 +1370,14 @@ def test_dt64arr_add_sub_DateOffset(self, box_with_array):
         tm.assert_equal(result, exp)
         tm.assert_equal(result2, exp)
 
+        # GH 26258
+        date = date_range(start='01 Jan 2014', end='01 Jan 2017', freq='AS')
```
Comment (jreback) [inline pandas/tests/arithmetic/test_datetime64.py]: can you also test with just pd.DateOffset(months=3) as well

File: pandas/tests/arithmetic/test_datetime64.py [hunk 3]
```diff
@@ -1370,6 +1370,14 @@ def test_dt64arr_add_sub_DateOffset(self, box_with_array):
...
+        offset = pd.DateOffset(months=3, days=10)
+        result = date + offset
```
Comment (jreback) [inline pandas/tests/arithmetic/test_datetime64.py]: also pls test for subtraction too

**Comment by makbigc** [comment]
@jreback Do you have anything else to implement? Please tell me.

**Comment by jreback** [comment]
thanks @makbigc keep em coming!

File: pandas/tests/arithmetic/test_datetime64.py [hunk 4]
```diff
@@ -1435,6 +1435,28 @@ def test_dt64arr_add_sub_offset_ndarray(self, tz_naive_fixture,
...
+         DatetimeIndex([Timestamp('2014-04-01'), Timestamp('2015-04-01'),
+                        Timestamp('2016-04-01'), Timestamp('2017-04-01')])),
+        ('__sub__', pd.DateOffset(months=3, days=10),
+         DatetimeIndex([Timestamp('2013-09-21'), Timestamp('2014-09-21'),
+                        Timestamp('2015-09-21'), Timestamp('2016-09-21')])),
+        ('__sub__', pd.DateOffset(months=3),
+         DatetimeIndex([Timestamp('2013-10-01'), Timestamp('2014-10-01'),
+                        Timestamp('2015-10-01'), Timestamp('2016-10-01')]))
+
+    ])
```
Comment (jbrockmendel) [inline pandas/tests/arithmetic/test_datetime64.py]: @makbigc I'm a little bit late to the party here, but can you do a follow-up to parametrize over `box` to make sure the add/sub ops work for Series/DataFrame?  Also tzaware and tz-naive.

**Pull request merged**